[reland] switch mean to use reduction linear

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97996

mean is actually a reduction linear formula if the final reduction
is partial sum (which currently is), so switching to use that instead

wanchaol committed:
[dtensor] switch mean to use reduction linear

mean is actually a reduction linear formula if the final reduction
is partial sum, so switching to use that instead

[ghstack-poisoned]

XilunWu reviewed (approved):
Thx for merge mean rule! One question.

XilunWu commented on torch/distributed/_tensor/ops/math_ops.py:
```diff
@@ -36,6 +36,9 @@ def default_reduction_rule(op_schema: OpSchema) -> OutputSharding:
+        aten.mean.out
     ]
 )
 def sum_rule(op_schema: OpSchema) -> OutputSharding:
```
since this rule is not only for `sum`, we can rename it to something like `linear_reduction_rule`.

wanchaol replied:
sure, will do it later

wanchaol commented:
@pytorchbot rebase

yifuwang reviewed (approved):
Awesome!

wanchaol committed:
Update on "[dtensor] switch mean to use reduction linear"


mean is actually a reduction linear formula if the final reduction
is partial sum (which currently is), so switching to use that instead

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
> @pytorchbot revert -m 'Sorry for reverting your PR, but it fails a test on CPU https://hud.pytorch.org/pytorch/pytorch/commit/1b323b313ce35c03583ece017f928079f4a86882' -c nosignal
> 
> I want to restore distributed CPU tests, but then it's still timeout frequently in turnk AFAICT

@huydhn I couldn't why see the CPU failure correlates to my change though..

huydhn commented:
> @huydhn I couldn't why see the CPU failure correlates to my change though..

I see. I make a guess because the test starts to fail with your change.  Let me run it with `ciflow/unstable`, and I will reland this if it pass.  I will also do a rebase just to avoid landrace.

huydhn commented:
@pytorchbot rebase

huydhn commented:
I think the same test also fails on the PR, thoughts?

wanchaol commented:
> I think the same test also fails on the PR, thoughts?

Oh now i see some error related, let me debug a bit and get back, thanks @huydhn!

wanchaol committed:
Update on "[dtensor] switch mean to use reduction linear"


mean is actually a reduction linear formula if the final reduction
is partial sum (which currently is), so switching to use that instead

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] switch mean to use reduction linear"


mean is actually a reduction linear formula if the final reduction
is partial sum (which currently is), so switching to use that instead

[ghstack-poisoned]

wanchaol commented:
> I think the same test also fails on the PR, thoughts?

@huydhn fixed the tests and unstable CI, landing. Thanks for detecting the issue!

wanchaol commented:
@pytorchbot merge
